Improve DHCP static leases management GUI

DL6ER opened the pull request:
# What does this implement/fix?

See title. This adds a easy accessible table that is kept in sync with the `<textarea>` which still allows for any advanced configuration.

Please see this movie for a demonstration:

[Screencast From 2025-07-15 12-41-22.webm](https://github.com/user-attachments/assets/7f800b49-1eb9-45e6-8d34-5f54c402536a)

You can test all of that even without having the DHCP server enabled. What will be saved in the end is ***always*** the content of said `<textarea>` which is getting stores as `dhcp.hosts` in `pihole.toml`.

---

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `development` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

DL6ER committed:
Add static DHCP lease editing table fully synced with the advanced textarea below

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Add copy from current to static DHCP leases button

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Add tooltips to the action buttons

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Remove unused code

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Bind data attributes instead of inlining them for security (to prevent any possibility for code injection)

Signed-off-by: DL6ER <dl6er@dl6er.de>

smithbill17 commented:
The video looks good to me, but unfortunately I'm not able to test this as I have now stopped using PiHole (partly due to this very issue).

However, two comments based on the video:
a) the 'advanced config' line shows *nothing* in the easy text grid - can't it still show MAC, IP & name but with an expansion arrow (or some kind of indicator) for the additional fields?
b) in PiHole v5, it didn't seem to matter if MACs were entered with hyphens ('-') rather than colons (':'). Is this the case with v6? And does it matter if IPs fall out-with the DHCP scope?

DL6ER committed:
Allow "-" as separator for MAC addresses

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
> a) the 'advanced config' line shows nothing in the easy text grid - can't it still show MAC, IP & name but with an expansion arrow (or some kind of indicator) for the additional fields?

Yes, this is expected and per design. Advanced config lines may include `set` and `id` and other identifiers and it'd be very complicated (= error-prone) to edit the "simple" parts of it without damaging the advanced config. Not trying to edit these lines but instead showing where they are in the text area (where the user initially created them) seems the most straightforward and most robust way

> b) in PiHole v5, it didn't seem to matter if MACs were entered with hyphens ('-') rather than colons (':'). Is this the case with v6? And does it matter if IPs fall out-with the DHCP scope?

I did add `-` as separators now. It doesn't matter if IP addresses are within the set DHCP scope.

DL6ER committed:
Add IPv6 support for static DHCP lease management

Signed-off-by: DL6ER <dl6er@dl6er.de>

yubiuser commented:
If something invalid is entered in the advanced box (e.g. `00:20:e0:3b:13:af,8.8.8.8.7,laptop`) the corresponding field in the table is empty (IP). Maybe we should show `invald` instead?

The live-editing is cool but only shown advanced > table, not the other way round. Is this by design?

> Advanced settings present in line 3

This could be hard to find if n is big. Could we show line numbers in the advanced text box?

DL6ER committed:
Add line numbers to dhcp.hosts textarea

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
> Could we show line numbers in the advanced text box?

Yes.

<img width="1011" height="778" alt="Screenshot From 2025-07-16 08-27-16" src="https://github.com/user-attachments/assets/245b870e-6474-4292-910f-783fc75694a9" />


> If something invalid is entered in the advanced box (e.g. 00:20:e0:3b:13:af,8.8.8.8.7,laptop) the corresponding field in the table is empty (IP). Maybe we should show invald instead?

The code is written such that it parses the same way `dnsmasq` parses this line. If the IP address is not valid, it is taken as hostname.

> The live-editing is cool but only shown advanced > table, not the other way round. Is this by design?

Yes. First, I implemented it both ways but as soon as you edit the IP address, e.g., `192.168.0.` (because you are editing the last digit), it'd immediately be dropped from the IP column as it is now invalid. This prevents further editing. Hence, I decided we let the user finish what they want and then click save.

yubiuser commented:
> If the IP address is not valid, it is taken as hostname.

Except when a hostname is also given
<img width="1724" height="680" alt="2025-07-16_09-09" src="https://github.com/user-attachments/assets/33d1c5cc-38e4-4b3a-ad5f-2e59539b7b75" />

rdwebdesign commented:
I tested using an invalid IP (`192.168.0.300`) and a hostname:

```
00:20:e0:3b:13:af,192.168.0.300,laptop
```

When saved, the invalid IP is ignored, resulting in `00:20:e0:3b:13:af,laptop`. 

I think this is the correct behavior, but no warning was returned to the user, to inform something was wrong and the entry was "fixed" to fit the valid input fields. Maybe we need a warning/alert for cases like this.

DL6ER committed:
Validate hwaddr and IP address while typing

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
Like so?

[Screencast From 2025-07-20 1.webm](https://github.com/user-attachments/assets/a92d0c6a-e2aa-4610-9133-d10f354e3db8)

DL6ER committed:
Reject saving if there are still validation errors in the table

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
[Screencast From 2025-07-20 20-39-42.webm](https://github.com/user-attachments/assets/c6c6ca7d-d82e-4749-8ca7-6063e1afaf89)

yubiuser commented:
I think the colors for invalid IP/MAC need a bit adjustment.
When using the light theme, the white color of the text has low contrast to the rosè background. When using a dark theme, the red color of the cell is almost invisible.

<img width="1202" height="222" alt="2025-07-22_15-38" src="https://github.com/user-attachments/assets/c4ba6122-1bb4-4b91-bc82-41bc325ff468" />

<img width="1112" height="224" alt="2025-07-22_15-38_1" src="https://github.com/user-attachments/assets/5ce0fefc-f532-465d-819b-229a93ad456b" />

rdwebdesign commented on style/pi-hole.css:
```diff
@@ -1609,3 +1612,13 @@ textarea.field-sizing-content {
   transition: opacity 200ms ease-in-out;
 }
+
+#StaticDHCPTable td.table-danger {
+  background-color: #ff000022 !important;
+  color: #fff !important;
+}
+
+#StaticDHCPTable td.table-success {
+  background-color: #74c70022 !important;
+  color: #fff !important;
+}
```
Quick suggestion just to improve the colors:

```suggestion
#StaticDHCPTable .table-danger {
  background-color: #d337;
}

#StaticDHCPTable .table-success {
  background-color: #2c24;
}
```

Still using semi-transparent colors, but now the contrast is good enough when using the original text color, so I think we don't need to change `color:`.

Also, we should use `!important` only in cases where we really need it.

DL6ER replied:
Oh, I did not see this before. I pushed non-transparent suggestions that should be the same on all variants, otherwise. Not overly pretty but it may suffice. Otherwise, we'll theme-dependent colors and this PR already became a lot more involved than I had hoped initially.

<img width="1014" height="205" alt="image" src="https://github.com/user-attachments/assets/0d3aec37-d10e-4f10-b1fd-12b4155a96c8" />

rdwebdesign committed:
Use background colors consistent with other tables on high contrast themes

Signed-off-by: RD WebDesign <github@rdwebdesign.com.br>

rdwebdesign commented:
I added one commit to change the high contrast themes. On these themes the colors are not as important as the contrast/readability.

rdwebdesign commented:
**Note:**

I was testing with DHCP server disabled, so I'm not sure if we need to fix this, but when you use the textearea to enter a line, if there are invalid values on this line, the values are not checked and the API accepts everything, including the invalid values.

The first line contains an invalid IP:
```
00:20:e0:3b:13:af,192.168.0.300,laptop,infinite
00:20:e0:3b:13:af,192.168.0.123,laptop
```

DL6ER committed:
Use non-transparent colors in the static DHCP table

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Better synchronize lines numbers and textarea

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
> I was testing with DHCP server disabled, so I'm not sure if we need to fix this, but when you use the textearea to enter a line, if there are invalid values on this line, the values are not checked and the API accepts everything, including the invalid values.

Yes, invalid `dhcp-host` lines are fine for `dnsmasq`. It just goes on when no useful content was found (like `0,0,0`). Invalid IP addresses, like in your example, will simply be interpreted as host names (just like the table does).

DL6ER committed:
Remove !important

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
The line numbers in the `<textarea>` have been improved:
[Screencast From 2025-07-24 16-41-59.webm](https://github.com/user-attachments/assets/0feadfc8-9596-406a-8c58-e85535558f41)

rdwebdesign commented:
The line numbers issue is fixed. Note: I'm not sure if I like the idea of removing the textarea resizing, but this is already doing what most users requested. We can think later about a solution for the resize.

I found another small issue: 

There is no hostname validation. If you enter a hostname containing spaces (or not allowed characters) the hostname will be saved. I tested using `laptop 2` and `laptop º 2`. Both were saved.
In this case I think we could add a validation function.
What do you think?

rdwebdesign commented:
Personally I think the text _"Please save this line before editing another or leaving the page, otherwise your changes will be lost."_ is a little misleading because even if the user saves the line (using the green button), the changes will still be lost if the user leaves the page without clicking on "Save & Apply".

I'm still unsure how to change this.

I had 2 different ideas:
1. we could use a check mark icon instead of a floppy disk icon and change the text to something like _"Please confirm the changes made on this line using the green button. Also, remember to click on "Save & Apply" before leaving the page, otherwise your changes will be lost."_.
2. we could try to remove the save/confirm button and automatically validate + save the changes (or show a warning) on `blur` (not tested yet).

yubiuser commented:
I agree with rd, it's not clear that the user needs to click "Save & Apply" after they already pressed the save icon

hemna commented:
please merge this and fix issues later.  pihole's dhcp is completely unusable as it is now.

yubiuser commented:
@DL6ER are you able to continue with this PR?

rk-h commented:
@DL6ER I'm not a programmer and I don't really know Git, but I would like to have this feature in my PiHole. Then I could take the Fritzbox out of service, because I only use it to assign a fixed DHCP address to a device after the initial installation.
vy 73

derchrisuk commented:
Can this be used with the latest pi-hole?
Docker Tag [2025.11.1](https://github.com/pi-hole/docker-pi-hole/releases/2025.11.1)Core [v6.3](https://github.com/pi-hole/pi-hole/releases/v6.3)FTL [v6.4.1](https://github.com/pi-hole/FTL/releases/v6.4.1)Web interface [v6.4](https://github.com/pi-hole/web/releases/v6.4)

derchrisuk commented:
@DL6ER can you update the branch with the latest changes?
Then I could test it with a local build.

PromoFaux commented:
I've rebased this branch on `development` and (hopefully) fixed the merge conflict.

> Docker Tag [2025.11.1](https://github.com/pi-hole/docker-pi-hole/releases/2025.11.1)Core [v6.3](https://github.com/pi-hole/pi-hole/releases/v6.3)FTL [v6.4.1](https://github.com/pi-hole/FTL/releases/v6.4.1)Web interface [v6.4](https://github.com/pi-hole/web/releases/v6.4)

If you would like to try this branch on docker, you can bake your own local image utilising this FTL branch by following these instructions:

https://docs.pi-hole.net/docker/build-image/

derchrisuk commented:
I have been running on this branch for last 2-3 days.
Works like a charm.
One change I would like to suggest, right now when you add additional options, it removes the entry from the gui view.
Not sure what else you can add, but i added the infinite option as per the examples.
Maybe this can be added as an option once you copy the lease.
Because the overview is much nicer than just the textarea box

DL6ER committed:
Merge development into new/simple-dhcp-static-leases and address review feedback

Resolve merge conflict in style/pi-hole.css (keep both StaticDHCPTable
styles and DNSSEC query log styles).

Address outstanding reviewer feedback:
- Change save button icon from floppy-disk to checkmark to clarify it
  confirms the row edit, not a final save
- Update hint text to mention "Save & Apply" is still needed
- Add hostname validation on the hostname cell (rejects spaces, commas,
  and other characters invalid in DNS names)

Signed-off-by: Dominik <dl6er@dl6er.de>

DL6ER commented:
I rebased on current `development` and resolved the merge conflict.

I also addressed the two outstanding review items:

1. Save button UX: Changed the row save icon from `floppy-disk` to a `checkmark`, and updated the hint text to: *"Please confirm changes using the green button, then click 'Save & Apply' before leaving the page."* This makes the two-step flow clearer — the green button confirms the row edit, "Save & Apply" persists to the config.

2. Hostname validation: Added input validation on the hostname cell, matching the existing MAC/IP validation pattern. Invalid hostnames (containing spaces, commas, or other characters not allowed in DNS names) are highlighted in red with a tooltip. Valid DNS hostname characters: letters, digits, hyphens, and dots.

rdwebdesign committed:
Fix regex for hostname validation in DHCP settings

When using the `v` flag, hyphens need to be escaped inside a character class.

Signed-off-by: RD WebDesign <github@rdwebdesign.com.br>

rdwebdesign commented:
This is the explanation for my commit:

From **_developer.mozilla.org_ - Regular expressions - Character class: [...], [^...]**: [v-mode character class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_class#v-mode_character_class):

> In addition to `]` and `\`, the following characters must be escaped in character classes if they represent literal characters: `(`, `)`, `[`, `{`, `}`, `/`, `-`, `|`.

rdwebdesign committed:
Add buttons to lines containing "advanced" settings

Signed-off-by: RD WebDesign <github@rdwebdesign.com.br>

rdwebdesign committed:
Make sure the .save-static-row button looks disabled when needed

Signed-off-by: RD WebDesign <github@rdwebdesign.com.br>

rdwebdesign commented:
I added buttons to lines containing advanced settings. It will allow to delete these lines and add lines after them, like we do on other rows. Only the "Save" button will be disabled:

![newDHCP_delete_button_advanced](https://github.com/user-attachments/assets/f9826ffb-0348-4e32-b5dc-cddd83c4b1e1)

rdwebdesign commented on scripts/js/settings-dhcp.js:
```diff
@@ -227,6 +237,312 @@ function processDHCPConfig() {
+    tbody.append(tr);
+  }
+
+  tbody.find(".save-static-row, .delete-static-row, .add-static-row").prop("disabled", false);
```
@DL6ER 

Why are you enabling all buttons here?
Are you sure this line is needed?

If the line is needed, we can enable the buttons, except Save buttons created with `.disabled` CSS class:

```suggestion
  tbody.find(".save-static-row:not(.disabled), .delete-static-row, .add-static-row").prop("disabled", false);
```

rdwebdesign committed:
Merge branch 'development' into new/simple-dhcp-static-leases

Signed-off-by: RD WebDesign <github@rdwebdesign.com.br>

rdwebdesign committed:
Fix prettier complaints

Fix regex adding `v` flag and fix line breaks to respect maximum lenght

Note: these issues were not previously reported because there was a merge
conflict. The prettier test was only executed after the merge conflict was
resolved.

Signed-off-by: RD WebDesign <github@rdwebdesign.com.br>

rdwebdesign commented on scripts/js/settings-dhcp.js:
```diff
@@ -227,6 +237,301 @@ function processDHCPConfig() {
+    tbody.append(tr);
+  }
+
+  tbody.find(".save-static-row, .delete-static-row, .add-static-row").prop("disabled", false);
```
This change was [already requested](https://github.com/pi-hole/web/pull/3565#discussion_r2999026840).

rdwebdesign commented on style/themes/high-contrast.css:
```diff
@@ -386,9 +386,12 @@ select:-webkit-autofill {
   background-color: #fcfcfc;
 }
 
+#StaticDHCPTable .table-danger,
 .blocked-row td {
   background-color: rgba(230, 160, 160, 0.3);
 }
+
+#StaticDHCPTable .table-success,
```
This will be fixed if my [previous suggestion to change `pi-hole.css`](https://github.com/pi-hole/web/pull/3565#discussion_r2223518774) is used.

We don't need the higher specificity on `pi-hole.css`. A better fix would be to remove `td` from `pi-hole.css`, as suggested before.

rdwebdesign commented on style/themes/high-contrast-dark.css:
```diff
@@ -914,9 +914,12 @@ input[type="password"]::-webkit-caps-lock-indicator {
 }
 
 /* Tables and Datatables */
+#StaticDHCPTable .table-danger,
 .blocked-row td {
   background-color: rgba(56, 0, 40, 0.35);
 }
+
+#StaticDHCPTable .table-success,
 .allowed-row td {
   background-color: rgba(0, 64, 64, 0.35);
```
This will be fixed if my [previous suggestion to change `pi-hole.css`](https://github.com/pi-hole/web/pull/3565#discussion_r2223518774) is used.

We don't need the higher specificity on `pi-hole.css`. A better fix would be to remove `td` from `pi-hole.css`, as suggested before.

yubiuser commented on scripts/js/settings-dhcp.js:
```diff
@@ -227,6 +237,301 @@ function processDHCPConfig() {
+  if (/id:|set:|tag:|ignore|lease_time|,\s*,/v.test(line)) return "advanced";
+
+  // Split the line by commas and trim whitespace
+  const parts = line.split(",").map(s => s.trim());
+
+  // If there are more than 3 parts or less than 2, it's considered advanced
+  if (parts.length > 3 || parts.length < 2) return "advanced";
+
+  // Check if first part is a valid MAC address
+  const haveMAC = parts.length > 0 && utils.validateMAC(parts[0]);
+  const hwaddr = haveMAC ? parts[0].trim() : "";
+
+  // Check if the first or second part is a valid IPv4 or IPv6 address
+  const hasSquareBrackets0 = parts[0][0] === "[" && parts[0].at(-1) === "]";
+  const ipv60 = hasSquareBrackets0 ? parts[0].slice(1, -1) : parts[0];
+  const hasSquareBrackets1 = parts.length > 1 && parts[1][0] === "[" && parts[1].at(-1) === "]";
+  const ipv61 = hasSquareBrackets1 ? parts[1].slice(1, -1) : parts.length > 1 ? parts[1] : "";
+  const firstIsValidIP = utils.validateIPv4(parts[0]) || utils.validateIPv6(ipv60);
+  const secondIsValidIP =
+    parts.length > 1 && (utils.validateIPv4(parts[1]) || utils.validateIPv6(ipv61));
+  const ipaddr = firstIsValidIP ? parts[0].trim() : secondIsValidIP ? parts[1].trim() : "";
+  const haveIP = ipaddr.length > 0;
+
+  // Check if the second or third part is a valid hostname
+  let hostname = "";
+  if (parts.length > 2 && parts[2].length > 0) hostname = parts[2].trim();
+  else if (parts.length > 1 && parts[1].length > 0 && (!haveIP || !haveMAC))
+    hostname = parts[1].trim();
```
This is a valid remark. `--dhcp-host` has only optional arguments so things can be quite fast 'advanced'
https://thekelleys.org.uk/dnsmasq/docs/dnsmasq-man.html

markusdd commented:
Hi all, was directed here based on issue #6593. Very encouraging to see the work on this. Have a couple suggestions, questions:

a) Would it make sense to also mark which of the static leases are active? This is usually a convenient way of finding out if a specific device is online and has accepted the static lease already. This is something many router UIs offer.
b) I am not sure if and how this is covered based on the videos posted, but is there now an easy way to make a dynamic lease into a static one again?
c) Many good router UIs also have the feature to be able to click either the hostname and/or ip to get to the web page of a given device if it exposes one. This is super handy especially when dealing with printers, scanners or smart home devices to configure them and find them in the network.

Other than that this looks very good and would v6 bring back to what v5 had already done nicely.

darkexplosiveqwx commented:
> Hi all, was directed here based on issue #6593. Very encouraging to see the work on this. Have a couple suggestions, questions:
> 
> a) Would it make sense to also mark which of the static leases are active? This is usually a convenient way of finding out if a specific device is online and has accepted the static lease already. This is something many router UIs offer. b) I am not sure if and how this is covered based on the videos posted, but is there now an easy way to make a dynamic lease into a static one again? c) Many good router UIs also have the feature to be able to click either the hostname and/or ip to get to the web page of a given device if it exposes one. This is super handy especially when dealing with printers, scanners or smart home devices to configure them and find them in the network.
> 
> Other than that this looks very good and would v6 bring back to what v5 had already done nicely.

Quick note:
Pi-hole is not a router.

I can already tell you that a) is not really possible. Pi-hole can detect when a device renews its DHCP lease, but most of the time the DHCP lease time is one week or longer. When Pi-hole only sees a sign of life from the device every week or so, I dont think that it would result in useful data. Routers can track whether a device is online or not by looking at the IP packets it sends to the internet, since the router is the gateway. This allows much more accurate online tracking. For the vast majority, Pi-hole is not the gateway device, and even if it were, Pi-hole is restricted to DNS and DHCP and not routing. 

Regarding c), it _could_ be possible, but I don't see it realistically being in Pi-hole. The goal of Pi-holes DHCP server has never been being a full-featured router.

Like Dan Schaper (Pi-hole co-founder) said on the Discourse feature request:
> I'll remind everyone that the DHCP server in Pi-hole was intended to be a fallback for people that had routers from their ISPs that locked out the ability to modify the DNS servers provided to the LAN clients.

b) is already possible with this PR (you can test it with `pihole checkout web new/simple-dhcp-static-leases` and go back to the stable release with `pihole checkout web master`, for docker see https://docs.pi-hole.net/docker/build-image/)

markusdd commented:
I'm aware of the scope. Nevertheless the DHCP function is a core functionality as the correct DNS'es are also being broadcast by it and I, and I assume many others, have decided to use pihole for exactly that purpose - because it makes you independent of whatever your router/modem/gateway/whatever provider thinks is good for you.
I could also do DHCP using my Unifi gateway, but that makes things much less easy to overlook in the end. Having dns entries and dhcp at one location just makes sense from an administration standpoint.

Happy to see b) addressed because that is actually most important.
The linking to the devices is of course nice to have a but it is a great convencience.

As for the lease activeness from a): Yes, that is kind of a heuristic. Probably also the least 'important' point.

As for c): I think that has no bearing on whether pihole is a router or not, it's just a useful convenience feature. It does not require any router capabilities.
